### ISSUE TITLE
torch.cuda.amp deployed with ddp model meets memory leak

### ISSUE THREAD

**Issue by Marimauer** [pytorch/pytorch#40073]
## 🐛 Bug

Hi @mcarilli, i have encountered some gpu memory leak problem when i deployed ddp model with torch.cuda.amp, which haven't meet one with nvidia.amp. It's a simple code, i ran it with 4 tesla V100, and the problem appeared very quickly within seven or eight steps. Belows shows my code.

## To Reproduce
1. model init with torch.nn.parallel.DistributedDataParallel 
2. with torch.cuda.amp.autocast():
           _output = model(_input)
           loss = criterion(_output, label)
    scaler.scale(loss).backward()
    scaler.step(optimizer)
    scaler.update()

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

 - PyTorch Version (e.g., 1.0): 1.6.0.dev20200615
 - OS (e.g., Linux): ubuntu 20
 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Build command you used (if compiling from source):
 - Python version: 3.6
 - CUDA/cuDNN version: 10.2
 - GPU models and configuration: 
 - Any other relevant information:

## Additional context

cc @ezyang @gchanan @zou3519 @mcarilli @ngimel @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @xush6528 @osalpekar

**Comment by ezyang** [comment]
Is there any chance we could get the full code that reproduces this error?

**Comment by rohan-varma** [comment]
I tried writing a repro over at https://gist.github.com/rohan-varma/d211d4801c41ace159dc614e7ae46222 with the rough steps given in the issue description, and ran a few variants of the script, but wasn't able to reproduce the OOM (the GPU mem usage stabilizes around 900 MiB).

**Comment by Marimauer** [comment]
> Is there any chance we could get the full code that reproduces this error?
def train():
  enumerate dataloader, then
  with torch.cuda.amp.autocast():
      _output = model(_input)
      loss = criterion(_output, label)
      scaler.scale(loss).backward()
      scaler.step(optimizer)
      scaler.update()

def nproc_main(opt):
   training init params
   model = torch.nn.parallel.DistributedDataParallel(model,device_ids=[nproc],find_unused_parameters=True)
   metric = torch.nn.parallel.DistributedDataParallel(metric,device_ids=[nproc],find_unused_parameters=True)
   train()
def main():
    torch.multiprocessing.spawn(nproc_main, nprocs=opt['nprocs'], args=(opt,))

**Comment by Marimauer** [comment]
> Is there any chance we could get the full code that reproduces this error?

def train():
	enumerate dataloader, then
	with torch.cuda.amp.autocast():
		_output = model(_input)
		loss = criterion(_output, label)
		scaler.scale(loss).backward()
		scaler.step(optimizer)
		scaler.update()

def nproc_main():
	training stuff init
	model = torch.nn.parallel.DistributedDataParallel(model,device_ids=[nproc],find_unused_parameters=True)
        train()

def main():
	torch.multiprocessing.spawn(nproc_main, nprocs=opt['nprocs'], args=(opt,))   

yes, these code is more detail. Run it with multiple gpus then you will reproduce the problem.

**Comment by Marimauer** [comment]
> I tried writing a repro over at https://gist.github.com/rohan-varma/d211d4801c41ace159dc614e7ae46222 with the rough steps given in the issue description, and ran a few variants of the script, but wasn't able to reproduce the OOM (the GPU mem usage stabilizes around 900 MiB).

Did you run it on a single device? I didn't meet one either when deploy on a device. Run it with multiple gpus then you will reproduce the problem.

**Comment by ngimel** [comment]
Can you please provide a runnable reproducible example, including the definition of your model. With your code, someone will have to spend a significant amount of time to make it runnable, and it does not provide nearly enough information.

**Comment by rohan-varma** [comment]
> Did you run it on a single device? I didn't meet one either when deploy on a device. Run it with multiple gpus then you will reproduce the problem.

Yes, I ran it with 2 GPUs and the repro spawns one process per GPU.

**Comment by Marimauer** [comment]
> Yes, I ran it with 2 GPUs and the repro spawns one process per GPU.

Thxs for your code, i have compared it with mine and found something different causing the memory leak problem, which is an additional loss. i will test my code and upload it later if it really cause.

**Comment by mcarilli** [comment]
Your code snippet is text formatted, are you running it with forward indented under autocast or not?
```
with torch.cuda.amp.autocast():
    _output = model(_input)
```
or
```
with torch.cuda.amp.autocast():
_output = model(_input)
```

**Comment by mcarilli** [comment]
> i have compared it with mine and found something different causing the memory leak problem, which is an additional loss. 

does that mean you fixed the memory leak, and that `torch.cuda.amp` and DDP are innocent?  can we close the issue?

**Comment by calclavia** [comment]
I'm running into a similar memory issue from a Pytorch 1.5 -> 1.6 upgrade, but haven't pinpointed a way to reproduce it in minimal code yet.

**Comment by mcarilli** [comment]
@calclavia It's not clear your issue is related, if you're still seeing it, please file a separate issue with a minimal reproducer.  nb: some previous users saw memory leaks on Turing GPUs that were [fixed with a cudnn update](https://github.com/pytorch/pytorch/issues/40911#issuecomment-691999671).

@Marimauer based on https://github.com/pytorch/pytorch/issues/40073#issuecomment-656303163 can we assume this is resolved?  Please reopen and clarify if it's not resolved.

**Comment by LifeBeyondExpectations** [comment]
Is there any problem when we set `find_unused_parameters=True` ??
I also encounter memory leak problem.
